[WIKI] Study Parameter - Identifiers (#4469)
aces/Loris: docs/wiki/01 - STUDY PARAMETERS SETUP/01 - Study Variables/01 - Identifiers.md

# Identifiers

## Overview
By default LORIS provides 2 different identifiers for each participant or subject: 

> - **CandID**, also known as the **DCCID**, is a unique randomized 6-digit numeric ID (e.g. '436792').  It is completely anonymous, and is assigned automatically by Loris upon participant registration. The CandID is not configurable.

Reviewer: ```suggestion
- **CandID**, also known as the **DCCID**, is a unique randomized 6-digit numeric ID (e.g. '436792') assigned automatically by LORIS upon participant registration. The CandID is not configurable.
```

I don't know if it really makes sense to refer to this as being anonymous. I kind of know what you mean but I think it's also confusing

Author: @johnsaigle I'm not the one who wrote it so if it makes sense to @driusan and/or @christinerogers I dont mind accepting the suggestion

Reviewer 2: I don't know what your question is.. it's removing one clause of the sentence, so it doesn't make it any more or less accurate.

> - **PSCID** (Project Study Center ID) is a unique configurable ID. It can be manually entered when registering a participant and may contain the site abbreviation followed by sequential or randomized characters (e.g. 'MTL0006')

Reviewer 2: ```suggestion
- **PSCID** (Project Study Center ID) is a unique configurable ID. It can be be setup to be either manually entered when registering a participant or automatically generated and usually contains the site abbreviation followed by sequential or randomized characters (e.g. 'MTL0006'), but its exact format is customizable.
```

The format and the generation of the **PSCID** must be configured by an admin at the setup stage.

## Configuration

### Front End
_not yet available_

### SQL
_not yet available_

### Adding identifiers from the config.xml file

Both the format and the generation of PSCIDs can be configured by the project in the `config.xml` file of the `%LORIS_ROOT%/project/` directory.

PSCIDs can be created for new subjects in one of 3 ways: *sequentially generated*, *manually entered*, or *randomly generated*.

1. ***sequential*** generates PSCIDs sequentially for each new participant registered. **(default)**

 ```xml
 <PSCID>
     <generation>sequential</generation> 
     <structure>
         <seq type="siteAbbrev"/>
         <seq type="numeric" minLength="4"/>
     </structure>
 </PSCID>
 ```

> 2. ***manual*** asks the user to enter the PSCID when registering a new participant.

Reviewer 3: This could mention that the structure will be converted in a perl regex when generation = user for validation... maybe

Author: @xlecours could you formulate something and propose it ?

Reviewer 2: I don't know why that would be relevant to someone setting up LORIS, it seems more like an implementation detail.

 ```xml
 <PSCID> 
     <generation>user</generation> 
     <structure>
         <seq type="alphanumeric" minLength="4"/> <!-- Ex: AAA1-->
     </structure>
 </PSCID>
 ```

3. ***random*** generates PSCIDs with a random numerical value for each new participant registered.

 ```xml
 <PSCID>
     <generation>random</generation> 
     <structure>
         <seq type="static">PROJ</seq>
         <seq type="siteAbbrev"/>
         <seq type="numeric" minLength="4"/>
     </structure>
 </PSCID>
 ```
 
 Options for the `type` element of the `<seq>` tag are:
  - `siteAbbrev`: A string value that will be used as a dynamic prefix. Value drawn from the `Alias` field of the `psc` table in the database.

>   - `static`: A string value that will be used as a fixed prefix. Value defined in the `config.xml` file.

Reviewer 2: ```suggestion
  - `static`: A string value that will be used as a fixed value. Value defined in the `config.xml` file.
```

AFAIK it doesn't need to be a prefix

Author: did you check the code or are you assuming ?

Author: https://github.com/aces/Loris/blob/major/php/libraries/SiteIDGenerator.php#L200

Reviewer 2: I'm just assuming because it would be stupid to have a system to declare the structure of a string, if all the individual components only worked as prefixes.

Reviewer: Yeah in effect this ends up always being a prefix. This is how you get the "DCC" for an ID like `DCC0123`.

>   - `numeric`: An integer value generated dynamically in accordance to the generation method defined. This type must be associated with a `minLength` element. 

Reviewer: For each of these it's also possible to specify an optional "maxLength" attritbute

Author: @johnsaigle "each of these" you mean including `siteAbbrev` or just `numeric`, `alphanumeric` and `alpha` ?

and is that combined with the `minLength` tag or one or the other ?

Reviewer: It includes all of them. For every generation method, you can also provide a minLength and maxLength.

You can use either minLength, maxLength, both or neither. They will both default to a value of 4.

Author: okai cool will add the info

- **PSCID** (Project Study Center ID) is a unique configurable ID. It can be be setup to be either manually entered when registering a participant or automatically generated and usually contains the site abbreviation followed by sequential or randomized characters (e.g. 'MTL0006'), but its exact format is customizable.

The format and the generation of the **PSCID** must be configured by an admin at the setup stage.

## Configuration

### Front End
_not yet available_

### SQL
_not yet available_

### Adding identifiers from the config.xml file

Both the format and the generation of PSCIDs can be configured by the project in the `config.xml` file of the `%LORIS_ROOT%/project/` directory.

PSCIDs can be created for new subjects in one of 3 ways: *sequentially generated*, *manually entered*, or *randomly generated*.

1. ***sequential*** generates PSCIDs sequentially for each new participant registered. **(default)**

 ```xml
 <PSCID>
     <generation>sequential</generation> 
     <structure>
         <seq type="siteAbbrev"/>

>          <seq type="numeric" minLength="4"/>

Reviewer 4: @ridz1208 Can you add 'minvalue' and 'maxvalue' details?

Author: If you know the details can you either formulate them and post it here or modify directly my branch??

Otherwise I'll have to go dig up the info in the code and it will take llonger

Reviewer 4: ```
     <min>0001</min>
     <max>9999</max> 
```
Minimum and maximum values can be specified for sequentially generated PSCIDs. By default sequence will start at the lowest possible values (i.e.: 0000, AAAA).

Author: @PapillonMcGill which types does it apply to ? alpha, numeric, all of them ?

Reviewer 4: all of them.

Author: I made changes on lines 32, 56 and 69
 - can you confirm the syntax is correct ?
 - also you mentioned in your comments for sequential generation, shouldn't it apply to random as well ?

Reviewer 4: Yes it does also apply to random

Author: changes are made

 ```xml
 <PSCID> 
     <generation>user</generation> 
     <structure>
         <seq type="alphanumeric" minLength="2"/> <!-- Ex: AAA1-->
     </structure>
 </PSCID>
 ```

3. ***random*** generates PSCIDs with a random numerical value for each new participant registered.

 ```xml
 <PSCID>
     <generation>random</generation> 
     <structure>
         <seq type="static">PROJ</seq>
         <seq type="siteAbbrev"/>
         <seq type="numeric" minLength="4" maxLength="4" min="0001" max="9999" />
     </structure>
 </PSCID>
 ```
 
 Options for the `type` element of the `<seq>` tag are:
  - `siteAbbrev`: A string value that will be used as a dynamic prefix. Value drawn from the `Alias` field of the `psc` table in the database.
  - `static`: A string value that will be used as a fixed prefix. Value defined in the `config.xml` file.
  - `numeric`: An integer value generated dynamically in accordance to the generation method defined.
  - `alphanumeric`: An alphanumeric string value generated dynamically in accordance to the generation method defined. 
  - `alpha`: An alphabetic string value generated dynamically in accordance to the generation method defined.
 

>   > Note: The types above can be associated with a `minLength` and/or a `maxLength` attribute. both `minLength` and `maxLength` default to `4` if not specified.

Reviewer 4: I think it would be better to specify in some way that the note apply to `numeric`, `alphanumeric` and `alpha` types only.  Something like ` The last 3 types above can ...`

Author: done

# Identifiers

## Overview

> By default LORIS provides 2 different identifiers for each participant or subject: 

Reviewer 5: should we mention the name `candidate` in this introductory line? (If we go ahead and adopt the term candidate later on. I just remember you mentioning something like that yesterday during the roadmap meeting ;). I have no preference).

Author: I'm gonna change it to candidate but just for reference we agreed and subject for the futur

> - **CandID**, also known as the **DCCID**, is a unique randomized 6-digit numeric ID (e.g. '436792') assigned automatically by LORIS upon participant registration. The CandID is not configurable.

Reviewer 6: To follow up on Cecile's comment ^^  about consistency, 
change "candidate" to "candidate/participant" since "participant" is used more consistently below.
I find it reads funny to say "there are 2 IDs per candidate" followed by "this ID happens upon participant registration"

Author: well I'm only talking about the CandId in this section... theres a section for PSCID below

for the candidate/participant, since we use candidate in LORIS I'll switch them over to candidate for now

Author: done

> - **PSCID** (Project Study Center ID) is a unique configurable ID. It can be be setup to be either manually entered when registering a participant or automatically generated and usually contains the site abbreviation followed by sequential or randomized characters (e.g. 'MTL0006'), but its exact format is customizable.

Reviewer 6: be be setup -> be set up

Author: done

The format and the generation of the **PSCID** must be configured by an admin at the setup stage.

## Configuration

### Configuration from the front end
_not yet available_

### Configuration from SQL
_not yet available_

### Configuration from the config.xml file

> Both the format and the generation of PSCIDs can be configured by the project in the `config.xml` file of the `%LORIS_ROOT%/project/` directory.

Reviewer 6: "by the project" :  either change to "by the admin" or remove.
* would also be nice to clarify: Is it configurable *per* project?

Author: done

PSCIDs can be created for new subjects in one of 3 ways: *sequentially generated*, *manually entered*, or *randomly generated*.

1. ***sequential*** generates PSCIDs sequentially for each new participant registered. **(default)**

 ```xml
 <PSCID>
     <generation>sequential</generation> 
     <structure>
         <seq type="siteAbbrev"/>

>          <seq type="numeric" minLength="2" maxLength="4" min="10" max="9999"/>

Reviewer 6: I think it would help to give an example. 
the min length is 2, maxLength 4, and the min value is 10
 does this mean MTL0001, MTL0002 is not possible?

Author: Im not sure of the behaviour so maybe I'll let @johnsaigle or @PapillonMcGill comment here.

for example, given that setting is the first ID generated `MTL10` or `MTL0010`?
is it wise to allow a min/max ? shouldnt we just define a fixed length ?
if `MTL10` is generated, does it also generate `MTL0010` when it gets to length 4 ? or is it smart enough to account for trailing 0s ??

Reviewer 4: PSCID will got from MTL10 to MTL9999 in that case.

Reviewer 4: Having a minlength and maxLength would create problem for generation of the next PSCID value. Next value is created using max(PSCID).
PSCID MTL100 would be less (in value) than MTL99. I believe that the min length should be equal to the length of max (4 in that case) to prevent that kind of problem. Unless the length of PSCID is taken into account in the query, which would make it not simple.

Reviewer 4: After verification, if the sequence is numerical, there is no problem (the prefix is removed before finding the max value). But if the sequence is alphanumerical (or alpha) then "Z" would be selected as max over "AA". Thus the code will try to create "AA" even if already existing.

@ridz1208 @driusan How do we modify? Do we require minLength and maxLength to be the same value (fix length) or modify IDGeneration code to fix this problem?

Author: For now I will stop modifying this document until we reach a solution. This is a bug not having anything to do with this doc.

My personal opinion, make it a fixed length for numerics, min/max lenghts dont make sense since trailing 00s dont change the value of numbers and ultimately its confusing to have both MTL99 and MTL0099 in one database

Reviewer 4: I would go for fixed length even for alpha or alphanumeric, not just numerical value, to prevent ordering problem.

Author: i can live with that

Reviewer: Good news! It already has to be a fixed length. See my review below.

Reviewer 2: >  since trailing 00s dont change the value of numbers 

???

Reviewer: @driusan A note from McGill HR re: salaries

     </structure>
 </PSCID>
 ```

2. ***manual*** asks the user to enter the PSCID when registering a new candidate.

 ```xml
 <PSCID> 
     <generation>user</generation> 
     <structure>

>          <seq type="alphanumeric" minLength="2"/> <!-- Ex: AAA1-->

Reviewer: ```suggestion
         <seq type="alphanumeric" length="2"/> <!-- Ex: AAA1-->
```

Reviewer 4: I would suggest to change the value to 4 (length("AAA1" == 4)!

Reviewer: It should actually be `AAA02` with a length of 2. `length` is the length of the ID, not the ID + the prefix.

Reviewer 4: confusing where which part is site ID and which is sequence number!

>          <seq type="alphanumeric" minLength="2"/> <!-- Ex: AAA1-->

Reviewer 4: ```suggestion
         <seq type="alphanumeric" length="2"/> <!-- Ex: A1-->
```

Reviewer: This suggestion isn't right. It should be `AAA02` given `AAA` as a prefix and `2` as length.

This comment should probably be fleshed out to something like:

```
 <PSCID> 
     <generation>user</generation> 
     <structure>
         <seq type="alphanumeric" minLength="4"/> <!-- Ex: MTL1A2B for a site MTL-->
```

Reviewer 4: better but should also have 
`<seq type="static">MTL</seq>`

     </structure>
 </PSCID>
 ```

3. ***random*** generates PSCIDs with a random numerical value for each new participant registered.

 ```xml
 <PSCID>
     <generation>random</generation> 
     <structure>
         <seq type="static">PROJ</seq>
         <seq type="siteAbbrev"/>

>          <seq type="numeric" length="4" min="1" max="9999"/> <!-- Ex: PROJMTL0001-->

Reviewer 4: Not really a random number.

Not a deal breaker but would be nice to have a real (or look-a-like) random portion

Author: @PapillonMcGill done

>          <seq type="numeric" minLength="4" maxLength="4" min="0001" max="9999" />

Reviewer: ```suggestion
         <seq type="numeric" length="4" min="1" max="9999" />
```

     </structure>
 </PSCID>
 ```
 
 Options for the `type` element of the `<seq>` tag are:
  - `siteAbbrev`: A string value that will be used as a dynamic prefix. Value drawn from the `Alias` field of the `psc` table in the database.
  - `static`: A string value that will be used as a fixed prefix. Value defined in the `config.xml` file.
  - `numeric`: An integer value generated dynamically in accordance to the generation method defined.
  - `alphanumeric`: An alphanumeric string value generated dynamically in accordance to the generation method defined. 
  - `alpha`: An alphabetic string value generated dynamically in accordance to the generation method defined.
 

>   > Note: The last 3 types above (`numeric`,`alphanumeric`,`alpha`) can be associated with a `minLength` and/or 

Reviewer: ```suggestion

```

>   a `maxLength` attribute. Both `minLength` and `maxLength` default to `4` if not specified.

Reviewer: ```suggestion
```

# Identifiers

## Overview
By default LORIS provides 2 different identifiers for each candidate: 

- **CandID**, also known as the **DCCID**, is a unique randomized 6-digit numeric ID (e.g. '436792') assigned automatically by LORIS upon the candidate's registration. The CandID is not configurable.

> - **PSCID** (Project Study Center ID) is a unique configurable ID. It can be set up to be either manually entered when registering a candidate or automatically generated and usually contains the site or project abbreviation followed by sequential or randomized characters (e.g. 'MTL0006'), but its exact format is customizable.

Reviewer 2: LORIS actually has 3 by default: ExternalID, which follows the same configuration mechanism as PSCID but is intended for creating distinct identifiers when distributing data externally.

Author: added please review the content

> The format and the generation of the **PSCID** must be configured by an admin at the setup stage.

Reviewer: ```suggestion
The format and the generation of the **PSCID** must be configured by an admin by editing the `config.xml` file.
```

I think "the setup stage" is a bit unclear

Reviewer: I'll approve after this change

Author: nice catch

Author: I will emphasize that when the setup section starts being added as well

## Configuration

### Configuration from the front end
_not yet available_

### Configuration from SQL
_not yet available_

### Configuration from the config.xml file

Both the format and the generation of PSCIDs can be configured by an administrator in the `config.xml` file of the `%LORIS_ROOT%/project/` directory. these settings are applied to any and all new candidates.

PSCIDs can be created for new subjects in one of 3 ways: *sequentially generated*, *manually entered*, or *randomly generated*.

1. ***sequential*** generates PSCIDs sequentially for each new candiddate registered. **(default)**

 ```xml
 <PSCID>
     <generation>sequential</generation> 
     <structure>
         <seq type="siteAbbrev"/>
         <seq type="numeric" length="4" min="10" max="9999"/>
     </structure>
 </PSCID>
 ```
 > Example PSCID generated: MTL1234
 > Where the site's alias is MTL

2. ***manual*** asks the user to enter the PSCID when registering a new candidate.

 ```xml
 <PSCID> 
     <generation>user</generation> 
     <structure>
         <seq type="alphanumeric" length="2"/>
     </structure>
 </PSCID>
 ```

>   > Example PSCID generated: A1

Reviewer 4: doesn't the user have to supply the PSCID when creating the candidate in that case?

Reviewer: Yeah it may be good to add a little note here saying `this value supplied by a user via the web interface`

Author: true but it will be validated, I'll change the wording

Author: done

3. ***random*** generates PSCIDs with a random numerical value for each new participant registered.

 ```xml
 <PSCID>
     <generation>random</generation> 
     <structure>

>          <seq type="static">PROJ</seq>

Reviewer: I feel like this example is a little ambiguous given that there's now also `projectAbbrev` option.
```suggestion
         <seq type="static">PREFIX</seq>
```

or maybe `STUDY`? Something that makes it more different from the Project Alias option.

The example comment on 62 would need to be updated too.

Author: done

# Identifiers

## Overview
By default LORIS provides 3 different identifiers for each candidate: 

- **CandID**, also known as the **DCCID**, is a unique randomized 6-digit numeric ID (e.g. '436792') assigned automatically by LORIS upon the candidate's registration. The CandID is not configurable.
- **PSCID** (Project Study Center ID) is a unique configurable ID. It can be set up to be either manually entered when registering a candidate or automatically generated and usually contains the site or project abbreviation followed by sequential or randomized characters (e.g. 'MTL0006'), but its exact format is customizable.

> - **ExternalID** is a unique configurable ID. It can be set up to be either manually entered when registering a candidate or automatically generated and it is generally completely de-identified (no site of project information incorporated) to be used for data dissemination.

Reviewer: ```suggestion
- **ExternalID** is a unique configurable ID. It can be set up to be either manually entered when registering a candidate or automatically generated and it is generally completely de-identified (no site or project information incorporated) to be used for data dissemination.
```

Reviewer: will approve after this

Reviewer 2: @ridz1208 ping

The format and the generation of the **PSCID** and **ExternalID** must be configured by an admin by editing the `config.xml` file.

## Configuration

### Configuration from the front end
_not yet available_

### Configuration from SQL
_not yet available_

### Configuration from the config.xml file

> Both the format and the generation of PSCIDs can be configured by an administrator in the `config.xml` file of the `%LORIS_ROOT%/project/` directory. these settings are applied to any and all new candidates.

Reviewer 2: ```suggestion
Both the format and the generation of PSCIDs can be configured by an administrator in the `config.xml`. These settings are applied to any and all new candidates.
```

You're assuming the location of config.xml. There are many places it could be located.

Reviewer: @driusan I think your suggestion should read 

"in the `config.xml` file. These "

or "in `config.xml`. These"

Reviewer 2: thanks, fixed the suggestion.

Author: you are correct

     <generation>random</generation> 
     <structure>
         <seq type="static">PROJ</seq>
         <seq type="projectAbbrev"/>

>          <seq type="numeric" length="4" min="1" max="9999"/> <!-- Ex: PROJMTL3994-->

Reviewer: > Ex: PROJMTL3994

This is a bit of a confusing example. Would the siteAbbrev in this case be `PROJMTL` or is this using both aliases?

It may be good to clarify this to be something like

> Ex: PROJMTL3994 given a Site with Alias $X and a Project with Alias $Y

Author: I just forgot to change the example after going from site to project. but ur suggestion is a good idea, I'll add it.

Side note, would it work here ? are we not limited to 1 prefix ?

Reviewer: I'm pretty sure only one is possible, yeah.

Author: @johnsaigle take a look again ?
